### ISSUE TITLE
Feature: add the name of the sender account into the from header

### ISSUE THREAD

**Issue by fejesjoco** [bruceg/nullmailer#52]
Many demons (eg. cron, zfs event demon) generate emails which is identical to this: `echo stuff | mail -s Something root`. The reason I'm using nullmailer is so that I receive these messages.

mail comes from another packages, like mailutils on Debian, and they call the MTA, nullmailer in this case (probably the sendmail binary).

mail, by default, doesn't generate a from header. The demons I use don't do it either (it would be clunky anyway). AFAIK this is appended by the MTA.

SMTP providers usually replace the email address in the from header to that of the authenticated user. Gmail certainly does this, my webhosting provider as well. I guess this is quite common.

The more complicated MTA's put the name of the sender account into the from header, not just the address. This is the name coming from /etc/passwd, or just defaults to the username.

nullmailer doesn't put a name there. And as said above, no other program does it either. And the from address is rewritten by the SMTP server. Which means that I don't know which account the mail is coming from.

If my assumptions above are wrong, I'm sorry. I think this is what's happening (and not happening) inside nullmailer but not 100% sure.

Would it be possible for nullmailer to put the account name into the from header? If doing a getpwnam(3) call and parsing the gecos field is a bit overkill here, then even just adding the username of the account would be a great help.

**Comment by fejesjoco** [comment]
I added a dummy patch, haven't tested it yet.

**Comment by bruceg** [comment]
I'm hesitant to parse gecos to derive a name. As far as I know there is no standard for the contents of the field, other just as a comment.

You say that "the from address is rewritten by the SMTP server". What SMTP server is rewriting the From header in the email?

**Comment by fejesjoco** [comment]
Yeah the gecos is quite complicated. This is what exim does: https://github.com/Exim/exim/blob/7952eef9f77899f36b23f1b9fa679f459cd52ffd/src/src/exim.c#L4654

But the loginname alone would be quite useful.

Gmail rewrites the email address in the From header to the authenticated login, but not the name. I assume people use Gmail a lot. Every mail sent from nullmailer through Gmail will have the same From header. Gmail saves the original From as X-Google-Original-From though, so the sender account can be seen in the full headers view, but it's quite hidden.

This only solves the problem of the login name though. If you have multiple boxes configured for the same SMTP account with nullmailer, you would probably need some Gmail aliasing tricks.

**Comment by fejesjoco** [comment]
I tried my proposed patch.

I ran the command with the original 2.1 binary:
`joco@hp:~$ echo x | mail -s test1 root`

Then I repeated the command with my patch in it.

The email comes from user joco and goes to user root. I'm using GMail.

![before](https://user-images.githubusercontent.com/8083452/35472213-49362490-036b-11e8-8d5a-e4427baca9c9.png)
![after](https://user-images.githubusercontent.com/8083452/35472214-4d944b48-036b-11e8-87dd-adc414cc5006.png)

What happens here is that Gmail rewrites the from address to the authenticated email address. So we can't see who the email comes from. But the to address is left as is. With the patch, the sender is easily distinguishable from the name, even though the address has been rewritten.

I find it ugly that they replace one address but not the other. But I assume Gmail is a common choice so this could help many people. We could also try other common SMTP providers to see what they do.

(The gecos is not even needed actually. It might still be nice to support it, and instead of trying to parse it, just run it through a user-supplied regexp for example.)
